docs/architecture.md

```diff
@@ -0,0 +1,47 @@
+# Architecture
+
+Or "How does Regal work?"
+
+As you might have [read](https://www.styra.com/blog/guarding-the-guardrails-introducing-regal-the-rego-linter/), Regal
+[uses Rego for linting Rego](https://www.styra.com/blog/linting-rego-with-rego/) — or rather, Rego policies turned into
+a JSON representation of their abstract syntax tree (AST).
+
+## High-level Overview
+
+When running Regal against a directory, like `regal lint my-policies/`, Regal does the following:
+
+- For each source file provided for linting, Regal parses the Rego into its AST representation. This AST representation
+  is then turned into JSON and provided as the **input** variable to the linter rules.
+- Each linter rule (and there are almost 40 of them at the time of writing this) uses the **input**, which contains
+  information such as the package name, what imports are used, and all the rules and the expressions they contain, to
+  determine whether the Rego policy linted contains any violations against the rule. An example could be a rule that
+  [forbids shadowing](https://docs.styra.com/regal/rules/bugs/rule-shadows-builtin) (i.e. using the same name as)
+  built-in functions and operators.
+- Since rule bodies aren’t necessarily flat, but may contain nested bodies of constructs such as
+  [comprehensions](https://www.openpolicyagent.org/docs/latest/policy-language/#comprehensions) or
+  [every](https://www.openpolicyagent.org/docs/latest/policy-language/#every-keyword) blocks, many linter rules need to
+  traverse all expressions in order to find what they are looking for. This is normally done with the help of the
+  built-in [walk](https://www.openpolicyagent.org/docs/latest/policy-reference/#graph) function.
+- Traversing huge AST structures — and some policies contain millions of AST nodes! — takes time. This isn’t noticeable
+  when linting a single file, but for some of the largest policy repositories out there, with several thousands of
+  policy files and tests, the cost may be prohibitive. To alleviate that, Regal evaluates as many files as it possibly
+  can in parallel.
```

Comment:
I would have described this as:

> To alleviate this, Regal is implemented to process files concurrently to minimise the impact of IO bound tasks and to make use of multiple cores when available.

> many files as it possibly can in parallel.

When I read this I thought we might get some nit picking from folk talking about goroutine scheduling yada yada.

Reply:
Sounds reasonable to me! I'll make the change as suggested.